libnetdata/json/json.c

```diff
@@ -0,0 +1,294 @@
+	if(!tokens) return NULL;
+
+	int ret = jsmn_parse(&parser, js, len, tokens, n);
+	while (ret == JSMN_ERROR_NOMEM) {
+		n *= 2;
+		jsmntok_t *new = reallocz(tokens, sizeof(jsmntok_t) * n);
+		if(!new) {
+			freez(tokens);
+			return NULL;
+		}
+		tokens = new;
+		ret = jsmn_parse(&parser, js, len, tokens, n);
+	}
+
+	if (ret == JSMN_ERROR_INVAL) {
+		error("JSON: Invalid json string.");
+		freez(tokens);
+		return NULL;
+	}
+	else if (ret == JSMN_ERROR_PART) {
+		error("JSON: Truncated JSON string.");
+		freez(tokens);
+		return NULL;
+	}
+
+	if(count) *count = (size_t)ret;
+
+	if(json_tokens < n) json_tokens = n;
+	return tokens;
+}
+
+int json_callback_print(JSON_ENTRY *e)
+{
+    BUFFER *wb=buffer_create(300);
+
+	buffer_sprintf(wb,"%s = ", e->name);
+    char txt[50];
+	switch(e->type) {
+		case JSON_OBJECT:
+			e->callback_function = json_callback_print;
+			buffer_strcat(wb,"OBJECT");
+			break;
+
+		case JSON_ARRAY:
+			e->callback_function = json_callback_print;
+			sprintf(txt,"ARRAY[%lu]", e->data.items);
+			buffer_strcat(wb, txt);
+			break;
+
+		case JSON_STRING:
+			buffer_strcat(wb, e->data.string);
+			break;
+
+		case JSON_NUMBER:
+            sprintf(txt,"%Lf", e->data.number);
+			buffer_strcat(wb,txt);
+
+			break;
+
+		case JSON_BOOLEAN:
+			buffer_strcat(wb, e->data.boolean?"TRUE":"FALSE");
+			break;
+
+		case JSON_NULL:
+			buffer_strcat(wb,"NULL");
+			break;
+	}
+    info("JSON: %s", buffer_tostring(wb));
+	buffer_free(wb);
+	return 0;
+}
+
+size_t json_walk_string(char *js, jsmntok_t *t, size_t start, JSON_ENTRY *e)
+{
+	char old = js[t[start].end];
+	js[t[start].end] = '\0';
+	e->original_string = &js[t[start].start];
+
+	e->type = JSON_STRING;
+	e->data.string = e->original_string;
+	if(e->callback_function) e->callback_function(e);
+	js[t[start].end] = old;
+	return 1;
+}
+
+size_t json_walk_primitive(char *js, jsmntok_t *t, size_t start, JSON_ENTRY *e)
+{
+	char old = js[t[start].end];
+	js[t[start].end] = '\0';
+	e->original_string = &js[t[start].start];
+
+	switch(e->original_string[0]) {
+		case '0': case '1': case '2': case '3': case '4': case '5': case '6': case '7':
+		case '8': case '9': case '-': case '.':
+			e->type = JSON_NUMBER;
+			e->data.number = strtold(e->original_string, NULL);
+			break;
+
+		case 't': case 'T':
+			e->type = JSON_BOOLEAN;
+			e->data.boolean = 1;
+			break;
+
+		case 'f': case 'F':
+			e->type = JSON_BOOLEAN;
+			e->data.boolean = 0;
+			break;
+
+		case 'n': case 'N':
+		default:
+			e->type = JSON_NULL;
+			break;
+	}
+	if(e->callback_function) e->callback_function(e);
+	js[t[start].end] = old;
+	return 1;
+}
+
+size_t json_walk_array(char *js, jsmntok_t *t, size_t nest, size_t start, JSON_ENTRY *e)
+{
+	JSON_ENTRY ne;
+
+	char old = js[t[start].end];
+	js[t[start].end] = '\0';
+	ne.original_string = &js[t[start].start];
+
+	memcpy(&ne, e, sizeof(JSON_ENTRY));
+	ne.type = JSON_ARRAY;
+	ne.data.items = t[start].size;
+	ne.callback_function = NULL;
+	strcpy(ne.name, "");
```

Comment:
`ne.name[0] = '\0'` has the same effect and it is way faster.

Reply:
Done